controllers/testtriggers/testtrigger_validator.go

```diff
@@ -0,0 +1,226 @@
+	}
+
+	if err := v.validateResourceSelector(t.Spec.ResourceSelector); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := v.validateTestSelector(ctx, t.Spec.TestSelector, t.Spec.Execution); err != nil {
+		allErrs = append(allErrs, err...)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return k8serrors.NewInvalid(
+		schema.GroupKind{
+			Group: testtriggerv1.GroupVersion.Group,
+			Kind:  "TestTrigger",
+		},
+		t.Name,
+		allErrs,
+	)
+}
+
+func (v *Validator) ValidateUpdate(ctx context.Context, old runtime.Object, new *testtriggerv1.TestTrigger) error {
+	var allErrs field.ErrorList
+
+	if err := v.validateResource(new.Spec.Resource); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := v.validateAction(new.Spec.Action); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := v.validateExecution(new.Spec.Execution); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := v.validateResourceSelector(new.Spec.ResourceSelector); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := v.validateTestSelector(ctx, new.Spec.TestSelector, new.Spec.Execution); err != nil {
+		allErrs = append(allErrs, err...)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return k8serrors.NewInvalid(
+		schema.GroupKind{
+			Group: testtriggerv1.GroupVersion.Group,
+			Kind:  testtriggerv1.Resource,
+		},
+		new.Name,
+		allErrs,
+	)
+}
+
+func (v *Validator) ValidateDelete(ctx context.Context, trigger *testtriggerv1.TestTrigger) error {
+	return nil
+}
+
+func (v *Validator) validateResourceSelector(resourceSelector testtriggerv1.TestTriggerSelector) *field.Error {
+	fld := field.NewPath("spec").Child("testSelector")
+	if err := v.validateSelector(fld, resourceSelector); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *Validator) validateTestSelector(
+	ctx context.Context,
+	testSelector testtriggerv1.TestTriggerSelector,
+	execution string,
+) field.ErrorList {
+	var allErrs field.ErrorList
+
+	fld := field.NewPath("spec").Child("testSelector")
+	if err := v.validateSelector(fld, testSelector); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if testSelector.Name != "" {
+		namespace := testSelector.Namespace
+		if namespace == "" {
+			namespace = testtrigger.DefaultNamespace
+		}
+		fld = fld.Child("name")
+		if err := v.getTestResource(ctx, fld, execution, namespace, testSelector.Name); err != nil {
+			allErrs = append(allErrs, err)
+		}
+	}
+
+	return allErrs
+}
+
+func (v *Validator) validateSelector(fld *field.Path, selector testtriggerv1.TestTriggerSelector) *field.Error {
+	if selector.Name != "" && len(selector.Labels) > 0 {
+		verr := field.Duplicate(fld, "either name or labels selector can be used")
+		return verr
+	}
+	return nil
+}
+
+func (v *Validator) getTestResource(
+	ctx context.Context,
+	fld *field.Path,
+	execution, namespace, name string,
+) *field.Error {
+	switch execution {
+	case testtrigger.ExecutionTestsuite:
+		var testsuite testsuitev2.TestSuite
```

Comment:
Should we have hard links there? If we use gitops approach, test or test suite can be created later

Reply:
I do not have a strong opinion, I'll remove it